File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-1.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number[][]}
     */
    // time - O(n), space - O(w)
    levelOrder(root) {
        if (!root) return [];
        const res = [];
        const queue = [root];
        let cur = 0;
        while (cur < queue.length) {
            const levelNodes = [];
            const levelSize = queue.length;
            while (cur < levelSize) {
                const node = queue[cur++];
                levelNodes.push(node.val);
                if (node.left) queue.push(node.left);
                if (node.right) queue.push(node.right);
            }
            res.push(levelNodes);
        }
        return res;
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    # time - O(n), space - O(h)
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root: return True

        self.unbalanced = -1
        def calcHeight(node):
            if not node: return 0
            left = calcHeight(node.left)
            right = calcHeight(node.right)
            if (left == self.unbalanced or right == self.unbalanced or abs(right - left) > 1): return self.unbalanced
            return 1 + max(left, right)

        return calcHeight(root) != self.unbalanced